src/Data/Nat/Binary/Subtraction.agda

```diff
@@ -0,0 +1,344 @@
+... | yes _  = refl
+... | no x≮y = contradiction x<y x≮y
+x≤y⇒x∸y≡0 {1+[2 x ]} {2[1+ y ]} (inj₁ (odd<even x≤y)) with x ≤? y
+... | yes _  = refl
+... | no x≰y = contradiction x≤y x≰y
+x≤y⇒x∸y≡0 {1+[2 _ ]} {1+[2 _ ]} (inj₁ (odd<odd x<y)) =
+  x≡0⇒double[x]≡0 (x≤y⇒x∸y≡0 (inj₁ x<y))
+x≤y⇒x∸y≡0 {1+[2 x ]} {1+[2 x ]} (inj₂ refl) =
+  x≡0⇒double[x]≡0 (x≤y⇒x∸y≡0 (≤-refl {x}))
+
+x∸x≡0 : ∀ x → x ∸ x ≡ 0ᵇ
+x∸x≡0 x = x≤y⇒x∸y≡0 (≤-refl {x})
+
+x>y⇒x∸y≢0 : ∀ {x y} → x > y → x ∸ y ≢ 0ᵇ
+x>y⇒x∸y≢0 {x'@(2[1+ x ])} {zero} _ = subst (_≢ 0ᵇ) (sym (x∸0≡x x')) 2[1+x]≢0
+x>y⇒x∸y≢0 {2[1+ _ ]} {2[1+ _ ]} (even<even y<x) = x≢0⇒double[x]≢0 (x>y⇒x∸y≢0 y<x)
+x>y⇒x∸y≢0 {2[1+ x ]} {1+[2 y ]} (odd<even y≤x) with x <? y
+... | no _    = 1+[2x]≢0 {x ∸ y}
+... | yes x<y = contradiction y≤x (<⇒≱ x<y)
+x>y⇒x∸y≢0 {x'@(1+[2 x ])} {y'@(2[1+ y ])} (even<odd y<x) =
+  subst (_≢ 0ᵇ) (pred[2d]≡x'∸y') pred[2d]≢0
+  where
+  -- d = x ∸ y
+  d≢0            = x>y⇒x∸y≢0 y<x
+  pred[2d]≡x'∸y' = sym (odd∸even-for> y<x)
+  1≤d            = x<y⇒suc[x]≤y (x≢0⇒x>0 d≢0)
+  2≤2d           = double-mono-≤ 1≤d
+  1≤pred[2d]     = pred-mono-≤ 2≤2d
+  0<pred[2d]     = suc[x]≤y⇒x<y 1≤pred[2d]
+  pred[2d]≢0     = >⇒≢ 0<pred[2d]
+x>y⇒x∸y≢0 {1+[2 _ ]} {1+[2 _ ]} (odd<odd y<x) = x≢0⇒double[x]≢0 (x>y⇒x∸y≢0 y<x)
+
+x∸y≡0⇒x≤y : ∀ {x y} → x ∸ y ≡ 0ᵇ → x ≤ y
+x∸y≡0⇒x≤y {x} {y} x∸y≡0 with x ≤? y
+... | yes leq = leq
+... | no x≰y  = contradiction x∸y≡0 (x>y⇒x∸y≢0 (≰⇒> x≰y))
+
+x<y⇒0<y∸x : ∀ {x y} → x < y → 0ᵇ < y ∸ x
+x<y⇒0<y∸x {x} {y} x<y with <-cmp (y ∸ x) 0ᵇ
+... | tri> _ _     gt = gt
+... | tri≈ _ y∸x≡0 _  = contradiction (x∸y≡0⇒x≤y y∸x≡0) (<⇒≱ x<y)
+... | tri< y∸x<0 _ _  = contradiction y∸x<0 x≮0
+
+---------------------------------------------------------------
+-- Properties of _∸_ and _+_ and _≤_
+
+x≥y⇒[x∸y]+y≡x : ∀ {x y} → x ≥ y → (x ∸ y) + y ≡ x
+x≥y⇒[x∸y]+y≡x {zero} {y} y≤0 = begin-equality
+  (0ᵇ ∸ y) + y     ≡⟨ cong₂ _+_ (cong (0ᵇ ∸_) y≡0) y≡0 ⟩
+  (0ᵇ ∸ 0ᵇ) + 0ᵇ   ≡⟨ +-identityʳ 0ᵇ ⟩
+  0ᵇ               ∎
+  where open ≤-Reasoning;  y≡0 = x≤0⇒x≡0 y≤0
+x≥y⇒[x∸y]+y≡x {2[1+ _ ]}                      (inj₁ 0<even)          = refl
+x≥y⇒[x∸y]+y≡x {x'@(2[1+ x ])} {y'@(2[1+ y ])} (inj₁ (even<even y<x)) =
+  begin-equality
+    x'∸y' + 2[1+ y ]                 ≡⟨ cong (x'∸y' +_) (2[1+_]-double-suc y) ⟩
+    (double x∸y) + (double (suc y))  ≡⟨ sym (double-distrib-+ x∸y (suc y)) ⟩
+    double (x∸y + suc y)             ≡⟨ cong double (+-comm x∸y (suc y)) ⟩
+    double (suc y + x∸y)             ≡⟨ cong double (suc-+ y x∸y) ⟩
+    double (suc (y + x∸y))           ≡⟨ cong (double ∘ suc) (+-comm y x∸y) ⟩
+    double (suc (x∸y + y))           ≡⟨ cong (double ∘ suc) $
+                                        x≥y⇒[x∸y]+y≡x (inj₁ y<x) ⟩
+    double (suc x)                   ≡⟨ sym (2[1+_]-double-suc x) ⟩
+    2[1+ x ]                         ∎
+  where open ≤-Reasoning;  x∸y = x ∸ y;  x'∸y' = x' ∸ y'
+x≥y⇒[x∸y]+y≡x {z@(2[1+ x ])} {2[1+ x ]} (inj₂ refl) = begin-equality
```

Comment:
This subcase can be replaced with `cong (_+ z) (x∸x≡0 z)`

Reply:
Done